File: 1-200/38-countAndSay/go/38-countAndSay.go
package main

import (
	"fmt"
	"strconv"
)

// 递归
func countAndSay(n int) string {
	if n == 1 {
		return "1"
	}
	s := countAndSay(n - 1)

	var ret string
	ch, count := s[0], 1
	for i := 1; i < len(s); i++ {
		if s[i] == ch {
			count++
			continue
		}

		ret += strconv.Itoa(count) + string(ch)
		count = 1
		ch = s[i]
	}

	return ret + strconv.Itoa(count) + string(ch)
}

// 非递归
func countAndSay2(n int) string {
	ret := "1"
	if n == 1 {
		return ret
	}

	countStr := func(s string) string {
		var ret string
		ch, count := s[0], 1
		for i := 1; i < len(s); i++ {
			if s[i] == ch {
				count++
				continue
			}

			ret += strconv.Itoa(count) + string(ch)
			count = 1
			ch = s[i]
		}
		return ret + strconv.Itoa(count) + string(ch)
	}

	for i := 2; i <= n; i++ {
		ret = countStr(ret)
	}

	return ret
}

// 题目 n 范围是  1 ～ 30
// 所以可以提前算出来存到 map。不过这出题意图，就不实现了

func main() {
	for i := 1; i < 10; i++ {
		fmt.Println("ret:", i, ",", countAndSay(i), countAndSay2(i))
	}
}

File: 1-200/38-countAndSay/rust/38-countAndSay.rs
struct Solution;

impl Solution {
    pub fn count_and_say(n: i32) -> String {
        if n == 1 {
            return "1".to_string();
        }

        let s = Solution::count_and_say(n - 1);
        let mut c = s.chars().nth(0).unwrap();
        let mut count = 1;
        let mut ret = String::new();
        for i in 1..s.len() {
            if s.chars().nth(i).unwrap() == c {
                count += 1;
                continue;
            }
            ret += &(count.to_string() + &c.to_string());
            count = 1;
            c = s.chars().nth(i).unwrap();
        }
        ret + &(count.to_string() + &c.to_string())
    }

    pub fn count_and_say2(n: i32) -> String {
        let mut ret: String = "1".to_string();
        if n == 1 {
            return ret;
        }

        let count_str = |s: String| -> String {
            let mut c = s.chars().nth(0).unwrap();
            let mut count = 1;
            let mut ret = String::new();
            for i in 1..s.len() {
                if s.chars().nth(i).unwrap() == c {
                    count += 1;
                    continue;
                }
                ret += &(count.to_string() + &c.to_string());
                count = 1;
                c = s.chars().nth(i).unwrap();
            }
            ret + &(count.to_string() + &c.to_string())
        };

        for _ in 1..n {
            ret = count_str(ret);
        }
        ret
    }
}

fn main() {
    for i in 1..=7 {
        println!("ret: {:}, {:?}", i, Solution::count_and_say(i));
        println!("ret: {:}, {:?}", i, Solution::count_and_say2(i));
    }
}
